token/program2/src/processor.rs

```diff
@@ -391,14 +438,78 @@ impl Processor {
         Ok(())
     }
 
+    /// Processes a [FreezeAccount](enum.TokenInstruction.html) instruction.
+    pub fn process_freeze_account(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        freeze: bool,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let source_account_info = next_account_info(account_info_iter)?;
+        let mint_info = next_account_info(account_info_iter)?;
+        let authority_info = next_account_info(account_info_iter)?;
+
+        let mut source_data = source_account_info.data.borrow_mut();
+        let source_account: &mut Account = state::unpack(&mut source_data)?;
+
+        if source_account.is_native {
+            return Err(TokenError::NativeNotSupported.into());
+        }
+        if mint_info.key != &source_account.mint {
+            return Err(TokenError::MintMismatch.into());
+        }
+        if freeze && source_account.is_initialized != AccountState::Initialized
```

Comment:
`!source_account.is_initialized()`?

Reply:
For ease of unpacking(checked), I have `Account::is_initialized()` return true if the account is Initialized or Frozen. So this needs to be more specific.
Open to other ideas, though.

Reply:
Why not have `is_initialized() return true if not `AccountState::Initialized`?, I think I might be missing something.

Reply:
Because then calling `unpack()` on a frozen Account will fail with `TokenError::UninitializedState`. That doesn't feel right to me for frozen Accounts, as they aren't Uninitialized.

Reply:
As discussed in DM, will update to `source_account.is_frozen()`